src/coreclr/hosts/corerun/corerun.cpp

```diff
@@ -131,20 +132,36 @@ static string_t build_tpa(const string_t& core_root, const string_t& core_librar
             string_t tmp = pal::build_file_list(dir, ext, [&](const char_t* file)
                 {
                     string_t file_local{ file };
+                    string_t file_name { file }; 
 
                     // Strip the extension.
                     if (pal::string_ends_with(file_local, ext_len, ext))
-                        file_local = file_local.substr(0, file_local.length() - ext_len);
+                        file_name = file_local.substr(0, file_local.length() - ext_len);
 
                     // Return true if the file is new.
-                    return name_set.insert(file_local).second;
+                    return name_path_map.insert({file_name, file_local}).second;
                 });
 
             // Add to the TPA.
             tpa_list << tmp;
         }
     }
 
+    tpa_items->assembly_count = static_cast<uint32_t>(name_path_map.size());
+    tpa_items->assembly_filepaths = new char*[tpa_items->assembly_count];
```

Comment:
Similarly, there is a lot of apps and libraries that depend on the existing documented key-value pairs in appcontext:

- https://learn.microsoft.com/en-us/dotnet/core/dependency-loading/default-probing
- https://stackoverflow.com/questions/75314798/is-there-a-net-framework-equivalent-to-cores-appcontext-getdatatrusted-platf

Are we going to break these or are these key/value pairs going to continue to work?

Reply:
The plan is not to break anything. I gather from @elinor-fung that the host_runtime_contract usage is intended to be gradual.